### PR TITLE
OSASINFRA-3363: ShiftStack CAPO integration

### PR DESCRIPTION
**Pull request by pierreprinetti**
Opening a PR so that Github keeps an history of the changes in this branch

### CODE REVIEW

**Comment by pierreprinetti** [comment]
We will have to live with that "Conflicting files" warning until @patrickdillon's PR is either rebased on or merged to master.

**Comment by patrickdillon** [comment]
working on adding an openstack capi job here: https://github.com/openshift/release/pull/48135

**Comment by pierreprinetti** [comment]
/hold
Depends on https://github.com/openshift/installer/pull/7967

**Comment by patrickdillon** [comment]
/test altinfra-e2e-openstack-capi-ovn

[impatient]

File: pkg/infrastructure/platform/platform.go [hunk 1]
```diff
@@ -62,6 +64,9 @@ func ProviderForPlatform(platform string, fg featuregates.FeatureGate) (infrastr
 	case powervstypes.Name:
 		return terraform.InitializeProvider(powervs.PlatformStages), nil
 	case openstacktypes.Name:
+		if fg.Enabled(configv1.FeatureGateClusterAPIInstall) {
+			return clusterapi.InitializeProvider(openstackcapi.Provider{}), nil
+		}
```
Comment (patrickdillon) [inline pkg/infrastructure/platform/platform.go]: Can you also add this to the platform_altinfra.go file?

I recently added a little background on the altinfra image in the enhancement: https://github.com/openshift/enhancements/blob/56e53e353b0201936b687e4f0d1ada3897ddf201/enhancements/installer/bootstrapping-clusters-with-capi-providers.md#test-plan

**Comment by MaysaMacedo** [comment]
The `altinfra-e2e-openstack-capi-ovn` job is [failing](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/7939/pull-ci-openshift-installer-master-altinfra-e2e-openstack-capi-ovn/1755242296367910912) because of the following issue being reported in the `openstackcluster` controller:
`"failed to reconcile external network: found 2 external networks, which should not happen"`
This is an expected issue that has been fixed upstream, once that is released we will bump the version used by the installer. Issue is tracked here https://issues.redhat.com/browse/OSASINFRA-3338.

**Comment by MaysaMacedo** [comment]
@patrickdillon FYI, we are trying to run the `altinfra-e2e-openstack-capi-ovn` on a different cloud to avoid the issue we hit in that job.

**Comment by mandre** [comment]
We need https://github.com/openshift/ci-tools/pull/3967 for testing

File: pkg/infrastructure/openstack/clusterapi/clusterapi.go [hunk 2]
```diff
@@ -89,6 +89,15 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
 		return fmt.Errorf("failed to get OSPCluster: %w", err)
 	}
 
+	bootstrapMachine := &capo.OpenStackMachine{}
```
Comment (mandre) [inline pkg/infrastructure/openstack/clusterapi/clusterapi.go:92]: The bootstrap node is not yet created when the `InfraReady` hook is called. You need to use `PostProvision` hook instead.

File: pkg/infrastructure/openstack/clusterapi/clusterapi.go [hunk 3]
```diff
@@ -116,6 +125,21 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
 		}
 	}
 
+	bootstrapPort, err := getPortForInstance(networkClient, *bootstrapMachine.Spec.InstanceID)
+	if err != nil {
+		return err
+	}
+
+	bootstrapFIP, err := createFIP(networkClient, "bootstrap", infraID, ospCluster.Status.Network.ID)
```
Comment (mandre) [inline pkg/infrastructure/openstack/clusterapi/clusterapi.go:133]: See my comment at https://github.com/shiftstack/installer/commit/585be0f0c86844ab8e69a53fa9dbdf957b1442c0#r138625202.

We should create and assign the FIP with a single API call.

**Comment by MaysaMacedo** [comment]
The issue we seen on the openstack job is also seen by other platforms:

`error running apply for configmap "openshift-cluster-api/cluster-api" (172 of 900): namespaces "openshift-cluster-api" not found`

All the operators are active, but installation does not succeed because of the above issue on CVO.

File: pkg/asset/machines/clusterapi.go [hunk 4]
```diff
@@ -154,7 +157,7 @@ func (c *ClusterAPI) Generate(dependencies asset.Parents) error {
 		}
 		c.FileList = append(c.FileList, awsMachines...)
 
-		// TODO(vincepri): The following code is almost duplicated from aws.AWSMachines.
+		// TODO(vincepri): The following code is almost duplicated from aws.GenerateMachines.
```
Comment (r4f4) [inline pkg/asset/machines/clusterapi.go]: Unrelated change

File: pkg/asset/machines/openstack/machines.go [hunk 5]
```diff
@@ -203,7 +203,7 @@ func generateProviderSpec(clusterID string, platform *openstack.Platform, mpool
 			Subnets: []machinev1alpha1.SubnetParam{
 				{
 					Filter: machinev1alpha1.SubnetFilter{
-						Name: fmt.Sprintf("%s-nodes", clusterID),
+						// Name: fmt.Sprintf("%s-nodes", clusterID),
```
Comment (r4f4) [inline pkg/asset/machines/openstack/machines.go]: Maybe could be removed?

File: pkg/infrastructure/openstack/clusterapi/clusterapi.go [hunk 6]
```diff
@@ -37,20 +40,77 @@ func (p Provider) ControlPlaneAvailable(in clusterapi.ControlPlaneAvailableInput
...
-		Name:      "CAPO test",
-		NetworkID: ospCluster.Status.Network.ID,
+	apiPort, err := createPort(networkClient, "api", in.InfraID, ospCluster.Status.Network.ID)
+	if err != nil {
+		return err
+	}
+	if in.InstallConfig.Config.OpenStack.APIFloatingIP != "" {
+		err = assignFIP(networkClient, in.InstallConfig.Config.OpenStack.APIFloatingIP, apiPort)
+		if err != nil {
+			return err
```
Comment (r4f4) [inline pkg/infrastructure/openstack/clusterapi/clusterapi.go]: Nit:
```go
if err := assignFIP(networkClient, in.InstallConfig.Config.OpenStack.APIFloatingIP, apiPort); err != nil {
    return err
}
```

File: pkg/infrastructure/openstack/clusterapi/clusterapi.go [hunk 7]
```diff
@@ -37,20 +40,77 @@ func (p Provider) ControlPlaneAvailable(in clusterapi.ControlPlaneAvailableInput
...
-	_, err = ports.Create(networkClient, createOtps).Extract()
+	ingressPort, err := createPort(networkClient, "ingress", in.InfraID, ospCluster.Status.Network.ID)
 	if err != nil {
 		return err
 	}
+	if in.InstallConfig.Config.OpenStack.IngressFloatingIP != "" {
+		err = assignFIP(networkClient, in.InstallConfig.Config.OpenStack.IngressFloatingIP, ingressPort)
+		if err != nil {
+			return err
+		}
```
Comment (r4f4) [inline pkg/infrastructure/openstack/clusterapi/clusterapi.go]: Nit:
```go
if err := ...; err != nil {
    return err
}
```

File: pkg/infrastructure/clusterapi/clusterapi.go [hunk 8]
```diff
@@ -286,6 +286,7 @@ func (i *InfraProvider) Provision(dir string, parents asset.Parents) ([]*asset.F
 	manifests := []client.Object{}
 	manifests = append(manifests, infraManifests...)
 	manifests = append(manifests, machineManifests...)
+
```
Comment (r4f4) [inline pkg/infrastructure/clusterapi/clusterapi.go]: Unrelated change.

File: vendor/github.com/gophercloud/gophercloud/openstack/imageservice/v2/imageimport/doc.go
Comment (r4f4) [inline vendor/github.com/gophercloud/gophercloud/openstack/imageservice/v2/imageimport/doc.go:1]: Vendor changes in their own commit.

File: pkg/infrastructure/openstack/infraready/floatingips.go [hunk 9]
```diff
@@ -45,7 +45,6 @@ func FloatingIPs(cluster *capo.OpenStackCluster, installConfig *installconfig.In
 	return nil
 }
 
-
```
Comment (r4f4) [inline pkg/infrastructure/openstack/infraready/floatingips.go]: Amend to the previous commit.

File: cluster-api/providers/openstack/go.mod
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply cluster-api/providers/openstack/go.mod:1]: done

File: cluster-api/providers/openstack/vendor/github.com/beorn7/perks/LICENSE
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply cluster-api/providers/openstack/vendor/github.com/beorn7/perks/LICENSE:1]: done

File: vendor/github.com/Azure/go-ansiterm/SECURITY.md
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply vendor/github.com/Azure/go-ansiterm/SECURITY.md:1]: done

File: pkg/infrastructure/openstack/clusterapi/clusterapi.go (hunk 6 above)
Comment (MaysaMacedo) [reply pkg/infrastructure/openstack/clusterapi/clusterapi.go]: done in the another commit

File: pkg/infrastructure/openstack/clusterapi/clusterapi.go (hunk 7 above)
Comment (MaysaMacedo) [reply pkg/infrastructure/openstack/clusterapi/clusterapi.go]: done in the another commit

File: pkg/infrastructure/clusterapi/clusterapi.go (hunk 8 above)
Comment (MaysaMacedo) [reply pkg/infrastructure/clusterapi/clusterapi.go]: done

File: vendor/github.com/gophercloud/gophercloud/openstack/imageservice/v2/imageimport/doc.go
Comment (MaysaMacedo) [reply vendor/github.com/gophercloud/gophercloud/openstack/imageservice/v2/imageimport/doc.go:1]: done

File: pkg/infrastructure/openstack/infraready/floatingips.go (hunk 9 above)
Comment (MaysaMacedo) [reply pkg/infrastructure/openstack/infraready/floatingips.go]: done

File: pkg/asset/machines/openstack/machines.go (hunk 5 above)
Comment (MaysaMacedo) [reply pkg/asset/machines/openstack/machines.go]: done

File: cluster-api/README.md [hunk 10]
```diff
@@ -2,4 +2,3 @@
 
 The OpenShift installer project leverages Cluster API and its infrastructure providers
 to create resources needed by an OpenShift cluster.
-
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply cluster-api/README.md]: done

**Review by r4f4 — COMMENTED** [review]
Code and commit structure LGTM. I'll leave it to @vincepri to review the capi bits.

**Comment by mandre** [comment]
The cluster install [was successful in the last run](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/7939/pull-ci-openshift-installer-master-altinfra-e2e-openstack-capi-ovn/1760281015797944320). However, because we're running this job on an environment that doesn't have octavia, a number of openstack-specific test fail. This isn't a big deal because we know these tests work against different environments, and we're going to move the job to one of these envs with Octavia as soon as we get a new version of CAPO.

File: pkg/asset/machines/openstack/openstackmachines.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+func GenerateMachines(clusterID string, config *types.InstallConfig, pool *types.MachinePool, osImage, role string) ([]*asset.RuntimeFile, error) {
+	if configPlatform := config.Platform.Name(); configPlatform != openstack.Name {
+		return nil, fmt.Errorf("non-OpenStack configuration: %q", configPlatform)
+	}
+	if poolPlatform := pool.Platform.Name(); poolPlatform != openstack.Name {
+		return nil, fmt.Errorf("non-OpenStack machine-pool: %q", poolPlatform)
+	}
+
+	mpool := pool.Platform.OpenStack
+	trunkSupport, err := checkNetworkExtensionAvailability(config.Platform.OpenStack.Cloud, "trunk", nil)
```
Comment (EmilienM) [inline pkg/asset/machines/openstack/openstackmachines.go]: for reviewers: it's good that we check for trunk availability this way, the install-config option for trunks has been deprecated for a long time.

File: pkg/asset/machines/openstack/openstackmachines.go [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+		var fixedIPs []capo.FixedIP
+		for _, fixedIP := range platform.ControlPlanePort.FixedIPs {
+			fixedIPs = append(fixedIPs, capo.FixedIP{
+				Subnet: &capo.SubnetFilter{
+					ID:   fixedIP.Subnet.ID,
+					Name: fixedIP.Subnet.Name,
+				}})
+		}
+		port.FixedIPs = fixedIPs
+		if len(addressPairs) > 0 {
```
Comment (EmilienM) [inline pkg/asset/machines/openstack/openstackmachines.go]: this is redundant in the else, we can probably move it out of the condition.

File: pkg/asset/machines/openstack/openstackmachines.go [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+	// FIXME: Uncomment when the server group rework merged
+	// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1779
+	// serverGroupName := clusterID + "-" + role
+	spec := capo.OpenStackMachineSpec{
+		CloudName: CloudName,
+		Flavor:    mpool.FlavorName,
+		// TODO(stephenfin): Create credentials
+		IdentityRef: &capo.OpenStackIdentityReference{
+			Kind: "Secret",
```
Comment (EmilienM) [inline pkg/asset/machines/openstack/openstackmachines.go]: This will be removed in CAPO when bumping to `v1beta1`.

File: pkg/asset/manifests/openstack/cluster.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+			// we (the installer) will create them, or the user will have
+			// pre-created them as part of a "Bring Your Own Network (BYON)"
+			// configuration
+			ManagedSecurityGroups:      false,
+			DisableAPIServerFloatingIP: true,
+			// TODO(stephenfin): update when we support dual-stack (there are
+			// potentially *two* IPs here)
+			APIServerFixedIP:  openstackInstallConfig.APIVIPs[0],
+			DNSNameservers:    openstackInstallConfig.ExternalDNS,
+			ExternalNetworkID: openstackInstallConfig.ExternalNetwork,
```
Comment (EmilienM) [inline pkg/asset/manifests/openstack/cluster.go]: This will be a network filter in CAPO when bumping to `v1beta1`.

File: pkg/asset/manifests/openstack/cluster.go [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+			Namespace: capiutils.Namespace,
+			Labels: map[string]string{
+				capi.ClusterNameLabel: clusterID.InfraID,
+			},
+		},
+		Spec: capo.OpenStackClusterSpec{
+			CloudName: CloudName,
+			// TODO(stephenfin): Create credentials
+			IdentityRef: &capo.OpenStackIdentityReference{
+				Kind: "Secret",
```
Comment (EmilienM) [inline pkg/asset/manifests/openstack/cluster.go]: This is removed when bumping to `v1beta1`.

**Comment by EmilienM** [comment]
I've reviewed (with my basic knowledge on the topic, but still a CAPO contributor), and manually tested it (due to the lack of a CI job), and it worked fine for me.
/lgtm

**Comment by r4f4** [comment]
/approve

/hold
Just adding a hold to give Vince some time to review.

**Comment by pierreprinetti** [comment]
@EmilienM 
* https://prow.ci.openshift.org/pr-history/?org=openshift&repo=installer&pr=7939
* the last altinfra-openstack run: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/7939/pull-ci-openshift-installer-master-altinfra-e2e-openstack-capi-ovn/1760281015797944320 Like @mandre says, the install was successful ( :tada: ) and just failed the loadbalancer tests

**Comment by MaysaMacedo** [comment]
Just amended one commit description
